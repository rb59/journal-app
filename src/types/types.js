export const types = {
    // Auth types
    login: '[Auth] Login',
    logout: '[Auth] Logout',
    // UI types
    setError: '[UI] Set Error',
    unsetError: '[UI] Unset Error',
    startLoading: '[UI] Start Loading',
    finishLoading: '[UI] Finish Loading',
    // Notes types
    addNew: '[Notes] New Note',
    setActive: '[Notes] Set Active Note',
    loadNotes: '[Notes] Load Notes',
    updateNote: '[Notes] Update Note',
    deleteNote: '[Notes] Delete Note',
    logoutCleaning: '[Notes] Logout Cleaning',
};
