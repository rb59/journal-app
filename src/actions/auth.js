import Swal from 'sweetalert2';
import { firebase, googleAuthProvider } from "../firebase/firebase-config";
import { types } from "../types/types";
import { finishLoading, startLoading } from "./ui";

export const startLogin = (email,password) => {
    return async(dispatch) => {
        dispatch( startLoading() );
        try {
            const {user:{uid,displayName}} = await firebase.auth().signInWithEmailAndPassword(email,password);  
            dispatch( login(uid,displayName) ); 
            dispatch( finishLoading() );
        } catch (e) {
            Swal.fire('Error',e.message,'error');
            dispatch( finishLoading() );
        }
    };
};

export const startRegister = (name,email,password) => {
    return async(dispatch) => {
        dispatch( startLoading() );
        try {
            const {user} = await firebase.auth().createUserWithEmailAndPassword(email,password);
            await user.updateProfile({displayName: name});
            const {uid,displayName} = user;
            dispatch( login(uid,displayName) );
            dispatch( finishLoading() );  
        } catch (e) {
            Swal.fire('Error',e.message,'error');
            dispatch( finishLoading() );  
        }
    };
};

export const startGoogleLogin = () => {
    return async(dispatch) => {
        try {
            const {user:{uid,displayName}} = await firebase.auth().signInWithPopup(googleAuthProvider);
            dispatch( login(uid,displayName) );
        } catch (e) {
            console.log(e);
        }
    };
};

export const login = (uid,displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    },
});

export const startLogout = () => {
    return async(dispatch) => {
        await firebase.auth().signOut();
        dispatch( logout() );
    };
}; 

export const logout = () => ({
    type: types.logout,
});