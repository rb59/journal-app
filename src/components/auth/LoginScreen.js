import React from 'react';
import {useDispatch, useSelector} from 'react-redux'
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLogin } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';
// cd
export const LoginScreen = () => {
    const dispatch = useDispatch();
    const { loading, msgError } = useSelector(state => state.ui);
    const [formValues,handleInputChange] = useForm({
        email: 'ronaldblanco0509@gmail.com',
        password: '123456',
    });

    const {email,password} = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch( startLogin(email,password) );
    }
    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    }

    return (
        <div>
            <h3 className="auth__title mb-5">Login</h3>
            <form onSubmit={handleLogin}>
                {
                    msgError && ( 
                    <div className="auth__alert-error" >
                        {msgError}
                    </div>
                    )
                }
                <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input mb-1"
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                    />
                <input 
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input mb-1"
                    value={password}
                    onChange={handleInputChange}
                />
                <button 
                    type="submit" 
                    className="btn btn-primary btn-block mb-1"
                    disabled={loading}
                    
                >
                    Login
                </button>
                <hr/>
                <div 
                    className="auth__social-networks"
                    onClick={handleGoogleLogin}
                >
                    <p>Login with Google</p>
                    <div 
                        className="google-btn"
                        >
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>
                <Link className="link" to="/auth/register">
                    Create new account
                </Link>
            </form>
        </div>
    );
};
