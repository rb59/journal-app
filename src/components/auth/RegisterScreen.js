import React from 'react';
import validator from 'validator';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { setError, unsetError } from '../../actions/ui';
import { useDispatch, useSelector } from 'react-redux';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector(({ ui }) => ui);
    const [formValues, handleInputChange] = useForm({
        name: 'Ronald',
        email: 'ronaldblanco0509@gmail.com',
        password: '123456',
        password2: '123456',
    });
    const { name, email, password, password2 } = formValues;

    const isFormValid = () => {
        if (name.trim().length === 0) {
            dispatch(setError('Name is required'));
            return false;
        } else if (!validator.isEmail(email)) {
            dispatch(setError('Email is not valid'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            dispatch(
                setError(
                    'Passwords should be at least 5 characters and match each other'
                )
            );
            return false;
        }
        dispatch(unsetError());
        return true;
    };
    const handleRegister = (e) => {
        e.preventDefault();
        if (isFormValid()) {
            dispatch(startRegister(name, email, password));
        }
    };
    return (
        <div>
            <h3 className="auth__title mb-5">Register</h3>
            <form
                onSubmit={handleRegister}
                className="animate__animated animate__fadeIn animate__faster"
            >
                {msgError && (
                    <div className="auth__alert-error">{msgError}</div>
                )}
                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input mb-1"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={name}
                />
                <input
                    type="email"
                    placeholder="Email"
                    name="email"
                    className="auth__input mb-1"
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={email}
                />
                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input mb-1"
                    onChange={handleInputChange}
                    value={password}
                />
                <input
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input mb-1"
                    onChange={handleInputChange}
                    value={password2}
                />
                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-1"
                    disabled={loading}
                >
                    Register
                </button>
                <hr className="mb-5" />

                <Link className="link" to="/auth/login">
                    Already registered?
                </Link>
            </form>
        </div>
    );
};
