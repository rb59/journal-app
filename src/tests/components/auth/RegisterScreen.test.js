import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';
// cd
const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initStore = {
    auth: {},
    ui: {
        loading: false,
        msgError: null,
    },
};

let store = mockStore(initStore);
//jest.mock('../../../actions/ui', () => ({
//setError: jest.fn(),
//}));
//store.dispatch = jest.fn()
describe('Pruebas en RegisterScreen', () => {
    beforeEach(() => {
        //store = mockStore(initStore);
        store.clearActions();
        jest.clearAllMocks();
    });

    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen />
            </MemoryRouter>
        </Provider>
    );

    test('debe hacer match con el snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('debe hacer el dispatch de la accion respectiva', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', {
            target: {
                name: 'email',
                value: '',
            },
        });
        wrapper.find('form').simulate('submit', {
            preventDefault() {},
        });
        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.setError,
            payload: 'Name is required',
        });
    });

    test('debe mostrar el mensaje de error', () => {
        const initStore = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email is not valid',
            },
        };
        const store = mockStore(initStore);
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(
            initStore.ui.msgError
        );
    });
});
