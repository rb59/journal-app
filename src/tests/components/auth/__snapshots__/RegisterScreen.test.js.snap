// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en RegisterScreen debe hacer match con el snapshot 1`] = `
<div>
  <h3
    className="auth__title mb-5"
  >
    Register
  </h3>
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input mb-1"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="Ronald"
    />
    <input
      autoComplete="off"
      className="auth__input mb-1"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value="ronaldblanco0509@gmail.com"
    />
    <input
      className="auth__input mb-1"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <input
      className="auth__input mb-1"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-1"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <hr
      className="mb-5"
    />
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>
</div>
`;
